tracer/transport.go

```diff
@@ -56,16 +54,23 @@ type httpTransport struct {
 	legacyTraceURL    string            // the legacy delivery URL for traces
 	serviceURL        string            // the delivery URL for services
 	legacyServiceURL  string            // the legacy delivery URL for services
-	pool              *encoderPool      // encoding allocates lot of buffers (which might then be resized) so we use a pool so they can be re-used
 	client            *http.Client      // the HTTP client used in the POST
 	headers           map[string]string // the Transport headers
 	compatibilityMode bool              // the Agent targets a legacy API for compatibility reasons
+
+	// [WARNING] We tried to reuse encoders thanks to a pool, but that led us to having race conditions.
+	// Indeed, when we send the encoder as the request body, the persistConn.writeLoop() goroutine
+	// can theoretically read the underlying buffer whereas the encoder has been returned to the pool.
+	// Since the underlying bytes.Buffer is not thread safe, this can make the app panicking.
+	// since this method will later on spawn a goroutine referencing this buffer.
+	// That's why we prefer the less performant yet SAFE implementation of allocating a new encoder every time we flush.
```

Comment:
One way to get the encoder overhead would be to just write a unit test that... only allocates an encoder. Totally agree that if the order of magnitude of this happening is "every 2sec", we'd be fine. Our current impl. tends to have big payloads flush rather than small calls, from a high-level view, is fine.

Reply:
I totally agree that we're good here. It's not a hot-path where if we have 1 millions of requests, we still allocate 1 encoder every 2 seconds. Good to keep this comment to explain why we undo that (premature) optimization.